File: src/main/java/org/virtuoso/escape/model/GameProjection.java
package org.virtuoso.escape.model;

import java.time.Duration;
import java.util.List;
import java.util.Optional;
import org.virtuoso.escape.model.account.Account;
import org.virtuoso.escape.model.account.AccountManager;
import org.virtuoso.escape.speech.SpeechPlayer;

/**
 * The game facade.
 *
 * @author Treasure
 */
public record GameProjection() {

    /**
     * Initialize and attempt login.
     *
     * @param username The username to attempt.
     * @param password The password to attempt.
     * @return {@code true} if the username-password combination was valid, otherwise {@code false}.
     */
    public boolean login(String username, String password) {
        Optional<Account> currentAccount = AccountManager.instance().login(username, password);
        return currentAccount.isPresent();
    }

    /**
     * Attempt to create an account.
     *
     * @param username The username to attempt.
     * @param password The password to attempt.
     * @return {@code true} if no user exists with this exact username-password combination, otherwise {@code false}.
     */
    public boolean createAccount(String username, String password) {
        Optional<Account> currentAccount = AccountManager.instance().newAccount(username, password);
        return currentAccount.isPresent();
    }

    /** Log the current user out and write data. */
    public void logout() {
        SpeechPlayer.instance().stopSoundbite();
        GameState.instance().write();
        AccountManager.instance().logout();
    }

    /**
     * The room the user is in.
     *
     * @return The room the user is in.
     */
    public Room currentRoom() {
        return GameState.instance().currentRoom();
    }

    /**
     * Change the current room.
     *
     * @param room The room to change to.
     */
    public void pickRoom(Room room) {
        GameState.instance().setCurrentRoom(room);
    }

    /**
     * The currently focused entity.
     *
     * @return The {@link Optional<Entity>} if the current entity is non-null, otherwise {@link Optional#empty()}.
     */
    public Optional<Entity> currentEntity() {
        return GameState.instance().currentEntity();
    }

    /**
     * The current floor.
     *
     * @return The current floor.
     */
    public Floor currentFloor() {
        return GameState.instance().currentFloor();
    }

    /**
     * Consume the current message.
     *
     * @return {@link Optional<String>} if the message is non-null, otherwise {@link Optional#empty()}.
     */
    public Optional<String> currentMessage() {
        return GameState.instance().currentMessage();
    }

    /**
     * The full inventory of the user.
     *
     * @return A list of items.
     */
    public List<Item> currentItems() {
        return GameState.instance().currentItems();
    }

    /**
     * Set the difficulty to {@code difficulty}
     *
     * @param difficulty The {@link Difficulty} to change to.
     */
    public void setDifficulty(Difficulty difficulty) {
        GameState.instance().setDifficulty(difficulty);
    }

    /**
     * Focus {@code entity}
     *
     * @param entity The {@link Entity} to focus.
     */
    public void pickEntity(Entity entity) {
        GameState.instance().pickEntity(entity);
    }

    /** Unfocus the current entity. */
    public void leaveEntity() {
        GameState.instance().leaveEntity();
    }

    /**
     * The time remaining on the countdown.
     *
     * @return The time remaining on the countdown.
     */
    public Duration time() {
        return GameState.instance().time();
    }

    /** Increments the {@code initialTime} by 1 minute if it is less than 2 hours. */
    public void incrementInitialTime() {
        GameState.instance().incrementInitialTime();
    }

    /**
     * Interact with the currently focused entity.
     *
     * @throws java.util.NoSuchElementException if no entity is focused.
     */
    public void interact() {
        currentEntity().orElseThrow().state().interact();
    }

    /**
     * Attack the currently focused entity.
     *
     * @throws java.util.NoSuchElementException if no entity is focused.
     */
    public void attack() {
        currentEntity().orElseThrow().state().attack();
    }

    /**
     * Inspect the currently focused entity.
     *
     * @throws java.util.NoSuchElementException if no entity is focused.
     */
    public void inspect() {
        currentEntity().orElseThrow().state().inspect();
    }

    /**
     * Speak to the current entity.
     *
     * @param input The input to give to the entity.
     * @throws java.util.NoSuchElementException if no entity is focused.
     */
    public void input(String input) {
        currentEntity().orElseThrow().state().takeInput(input);
    }

    /**
     * The capabilities of the current entity -- whether it supports an action.
     *
     * @return The capabilities of the entity.
     */
    public EntityState.Capabilities capabilities() {
        return currentEntity().orElseThrow().state().capabilities();
    }

    /**
     * Whether the game is ended.
     *
     * @return {@code true} if the game has ended, otherwise {@code false}.
     */
    public boolean isEnded() {
        return GameState.instance().isEnded();
    }
}

File: src/commonMain/kotlin/org.virtuoso.escape.model/GameProjection.kt
@file:OptIn(ExperimentalUuidApi::class)

package org.virtuoso.escape.model

import org.virtuoso.escape.model.account.Account
import kotlin.time.Duration
import kotlin.uuid.ExperimentalUuidApi

/**
 * The game facade.
 *
 * @author Treasure
 */
class GameProjection(fileReader: ((String) -> String), fileWriter: (String, String) -> Unit) {
  init {
    Data.FILE_READER = fileReader
    Data.FILE_WRITER = fileWriter
  }

  val accounts = Data.loadAccounts()
  val gamestates = Data.loadGameStates()
  lateinit var state: GameState
  lateinit var account: Account
  val language = Data.loadGameLanguage()!!

  /**
   * Initialize and attempt login.
   *
   * @param username The username to attempt.
   * @param password The password to attempt.
   * @return `true` if the username-password combination was valid, otherwise `false`.
   */
  fun login(username: String, password: String): Boolean {
    Account.login(username, password, accounts).let { acc ->
      account = acc
      state = gamestates[account.id]!!
      state.language = language
      state.floor.rooms
          .flatMap { it.entities }
          .map { state.currentEntityStates[it.id]?.let(it::swapState) }
    }
    return true
  }

  /**
   * Attempt to create an account.
   *
   * @param username The username to attempt.
   * @param password The password to attempt.
   * @return `true` if no user exists with this exact username-password combination, otherwise
   *   `false`.
   */
  fun createAccount(username: String, password: String): Boolean {
    Account.newAccount(username, password, this.accounts).let {
      account = it
      this.state = GameState()
      state.language = language
      return true
    }
  }

  /** Log the current user out and write data. */
  @OptIn(ExperimentalUuidApi::class)
  fun logout() {
    Data.writeAccount(account, accounts)
    Data.writeGameState(state, account, gamestates)
  }

  /**
   * The room the user is in.
   *
   * @return The room the user is in.
   */
  fun currentRoom(): Room {
    return this.state.room
  }

  /**
   * Change the current room.
   *
   * @param room The room to change to.
   */
  fun pickRoom(room: Room) {
    this.state.room = room
  }

  /**
   * The currently focused entity.
   *
   * @return The currently selected [Entity] or null.
   */
  fun currentEntity(): Entity? {
    return this.state.entity
  }

  /**
   * The current floor.
   *
   * @return The current floor.
   */
  fun currentFloor(): Floor {
    return this.state.floor
  }

  /**
   * Consume the current message.
   *
   * @return the message or null.
   */
  fun currentMessage(): String? {
    return this.state.message
  }

  /**
   * The full inventory of the user.
   *
   * @return A list of items.
   */
  fun currentItems(): MutableList<Item> {
    return this.state.items.toMutableList()
  }

  /**
   * Set the difficulty to `difficulty`
   *
   * @param difficulty The [Difficulty] to change to.
   */
  fun setDifficulty(difficulty: Difficulty) {
    this.state.difficulty = difficulty
  }

  /**
   * Focus `entity`
   *
   * @param entity The [Entity] to focus.
   */
  fun pickEntity(entity: Entity?) {
    this.state.entity = (entity)
  }

  /** Unfocus the current entity. */
  fun leaveEntity() {
    this.state.leaveEntity()
  }

  /**
   * The time remaining on the countdown.
   *
   * @return The time remaining on the countdown.
   */
  fun time(): Duration {
    return this.state.countdown()
  }

  /** Reset time remaining on the countdown. */
  fun resetTimer() {
    this.state.resetTimer()
  }

  /** Increments the `initialTime` by 1 minute if it is less than 2 hours. */
  fun incrementInitialTime() {
    this.state.incrementInitialTime()
  }

  /**
   * Interact with the currently focused entity.
   *
   * @throws [NullPointerException] if no entity is focused.
   */
  fun interact() {
    currentEntity()!!.state().interact(state)
  }

  /**
   * Attack the currently focused entity.
   *
   * @throws [NullPointerException] if no entity is focused.
   */
  fun attack() {
    currentEntity()!!.state().attack(state)
  }

  /**
   * Inspect the currently focused entity.
   *
   * @throws [NullPointerException] if no entity is focused.
   */
  fun inspect() {
    currentEntity()!!.state().inspect(state)
  }

  /**
   * Speak to the current entity.
   *
   * @param input The input to give to the entity.
   * @throws [NullPointerException] if no entity is focused.
   */
  fun input(input: String) {
    currentEntity()!!.state().takeInput(input, state)
  }

  /**
   * The capabilities of the current entity -- whether it supports an action.
   *
   * @return The capabilities of the entity.
   */
  fun capabilities(): EntityState.Capabilities {
    return currentEntity()!!.state().capabilities
  }

  val isEnded: Boolean
    /**
     * Whether the game is ended.
     *
     * @return `true` if the game has ended, otherwise `false`.
     */
    get() = this.state.isEnded
}
